src/mbgl/text/collision_tile.cpp

```diff
@@ -149,48 +167,71 @@ Box CollisionTile::getTreeBox(const Point<float>& anchor, const CollisionBox& bo
-    if (queryGeometry.empty()) return result;
+    if (queryGeometry.empty() || (tree.empty() && ignoredTree.empty())) {
+        return result;
+    }
 
+    // Account for the rounding done when updating symbol shader variables.
+    float zoom = util::log2(scale);
+    zoom = std::ceil(zoom * 10.0f) / 10.0f;
+    scale = std::pow(2.0f, zoom);
+
+    // Generate a rotated geometry out of the original query geometry.
+    // Scale has already been handled by the prior conversions.
+    BGMPolygon polygon;
+    for (const auto& point : queryGeometry) {
+        auto rotated = util::matrixMultiply(rotationMatrix, convertPoint<float>(point));
+        boost::geometry::append(polygon, CollisionPoint { rotated.x, rotated.y });
+    }
+
+    // Predicate for ruling out already seen features.
     std::unordered_map<std::string, std::unordered_set<std::size_t>> sourceLayerFeatures;
+    auto seenFeature = [&] (const CollisionTreeBox& treeBox) -> bool {
+        const IndexedSubfeature& feature = std::get<2>(treeBox);
+        const auto& seenFeatures = sourceLayerFeatures[feature.sourceLayerName];
+        return seenFeatures.find(feature.index) == seenFeatures.end();
+    };
 
-    mapbox::geometry::multi_point<float> rotatedPoints {};
-    rotatedPoints.reserve(queryGeometry.size());
-    std::transform(queryGeometry.cbegin(), queryGeometry.cend(), std::back_inserter(rotatedPoints),
-                   [&](const auto& c) { return util::matrixMultiply(rotationMatrix, convertPoint<float>(c)); });
-    const auto box = mapbox::geometry::envelope(rotatedPoints);
+    // Check if feature is rendered (collision free) at current scale.
+    auto visibleAtScale = [&] (const CollisionTreeBox& treeBox) -> bool {
+        const CollisionBox& box = std::get<1>(treeBox);
+        return scale >= box.placementScale && scale <= box.maxScale;
+    };
 
-    const auto& anchor = box.min;
-    CollisionBox queryBox(anchor, 0, 0, box.max.x - box.min.x, box.max.y - box.min.y, scale);
-    auto predicates = bgi::intersects(getTreeBox(anchor, queryBox));
+    // Check if query polygon intersects with the feature box at current scale.
+    auto intersectsAtScale = [&] (const CollisionTreeBox& treeBox) -> bool {
+        const Box& box = std::get<0>(treeBox);
+        const auto anchor = bg::return_centroid<CollisionPoint>(box);
+        // Unscale the tree box using its center as transform origin.
+        auto unscaledBox = Box {
+            CollisionPoint {
+                anchor.get<0>() - ((anchor.get<0>() - box.min_corner().get<0>()) / scale),
+                anchor.get<1>() - ((anchor.get<1>() - box.min_corner().get<1>()) / scale),
+            },
+            CollisionPoint {
+                anchor.get<0>() + ((box.max_corner().get<0>() - anchor.get<0>()) / scale),
+                anchor.get<1>() + ((box.max_corner().get<1>() - anchor.get<1>()) / scale),
+            }
+        };
+        return bg::intersects(polygon, unscaledBox);
```

Comment:
We don't have boost in -js so we'll need to write the intersection test ourselves there. It might be worth using our own version in -native as well just so that it matches -js exactly. I think it's already mostly written. See [polygonIntersectsMultipolygon](https://github.com/mapbox/mapbox-gl-native/blob/b774aa4bb0a3eaf76aeaabb9bc6ea308db09706c/src/mbgl/util/intersection_tests.cpp#L117-L137). If queryGeometry is just a polygon we might want to add a non-multi-polygon version.